test/MSTest.ForTesting.TestAdapter/TestPlatformAdapter.cs

```diff
@@ -0,0 +1,219 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Reflection;
+using System.Threading;
+
+using Microsoft.VisualStudio.TestPlatform.ObjectModel;
+using Microsoft.VisualStudio.TestPlatform.ObjectModel.Adapter;
+using Microsoft.VisualStudio.TestPlatform.ObjectModel.Logging;
+
+namespace MSTest.ForTesting.TestAdapter;
+
+[DefaultExecutorUri(Constants.ExecutorUri)]
+[ExtensionUri(Constants.ExecutorUri)]
+internal sealed class TestPlatformAdapter : ITestDiscoverer, ITestExecutor
+{
+    private CancellationTokenSource _testRunCancellationTokenSource;
+
+    /// <inheritdoc/>
+    public void DiscoverTests(IEnumerable<string> sources, IDiscoveryContext discoveryContext, IMessageLogger logger, ITestCaseDiscoverySink discoverySink)
+    {
+        foreach (var testCase in DiscoverTests(sources, logger))
+        {
+            discoverySink.SendTestCase(testCase);
+        }
+    }
+
+    public void Cancel()
+    {
+        _testRunCancellationTokenSource.Cancel();
+    }
+
+    public void RunTests(IEnumerable<TestCase> tests, IRunContext runContext, IFrameworkHandle frameworkHandle)
+    {
+        if (!Debugger.IsAttached) Debugger.Launch();
+        _testRunCancellationTokenSource = new CancellationTokenSource();
```

Comment:
maybe creates above...dunno if cancel can be called also for discovery, or `?` in the cancel